core/state/database.go

```diff
@@ -250,6 +252,118 @@ func (tds *TrieDbState) Copy() *TrieDbState {
 	return &cpy
 }
 
+// PutTombstoneForDeletedAccount - placing tombstone only if given account has storage in database
+func PutTombstoneForDeletedAccount(db ethdb.MinDatabase, addrHash []byte) error {
+	if len(addrHash) != common.HashLength {
+		return nil
+	}
+
+	var boltDb *bolt.DB
+	if hasKV, ok := db.(ethdb.HasKV); ok {
+		boltDb = hasKV.KV()
+	} else {
+		return fmt.Errorf("only Bolt supported yet, given: %T", db)
+	}
+
+	hasIH := false
+	if err := boltDb.View(func(tx *bolt.Tx) error {
+		k, _ := tx.Bucket(dbutils.IntermediateTrieHashBucket).Cursor().Seek(addrHash)
+		if !bytes.HasPrefix(k, addrHash) {
+			k = nil
+		}
+		if k != nil {
+			hasIH = true
+		}
+
+		return nil
+	}); err != nil {
+		return err
+	}
+
+	if !hasIH {
+		return nil
+	}
+
+	return db.Put(dbutils.IntermediateTrieHashBucket, common.CopyBytes(addrHash), []byte{})
+}
+
+func ClearTombstonesForNewStorage(db ethdb.MinDatabase, storageKeyNoInc []byte) error {
+	var boltDb *bolt.DB
+	if hasKV, ok := db.(ethdb.HasKV); ok {
+		boltDb = hasKV.KV()
+	} else {
+		return fmt.Errorf("only Bolt supported yet, given: %T", db)
+	}
+
+	var toPut [][]byte
+	toDelete := map[string]struct{}{}
```

Comment:
Does `toDelete` need to be a map? Do you expect duplicated entries? If not, can it be the same type as `toPut`?